src/BuiltInTools/dotnet-watch/Browser/BrowserConnector.cs

```diff
@@ -10,22 +12,96 @@ namespace Microsoft.DotNet.Watcher.Tools
     internal sealed partial class BrowserConnector(DotNetWatchContext context) : IAsyncDisposable
     {
         // This needs to be in sync with the version BrowserRefreshMiddleware is compiled against.
-        private static readonly Version s_minimumSupportedVersion = new(6, 0);
+        private static readonly Version s_minimumSupportedVersion = Versions.Version6_0;
 
         private static readonly Regex s_nowListeningRegex = s_nowListeningOnRegex();
 
         [GeneratedRegex(@"Now listening on: (?<url>.*)\s*$", RegexOptions.Compiled)]
         private static partial Regex s_nowListeningOnRegex();
 
-        private bool _attemptedBrowserLaunch = false;
-        public BrowserRefreshServer? RefreshServer { get; private set; }
+        private readonly object _serversGuard = new();
+        private readonly Dictionary<ProjectGraphNode, BrowserRefreshServer?> _servers = [];
+
+        // interlocked
+        private ImmutableHashSet<ProjectGraphNode> _browserLaunchAttempted = [];
+
+        public async ValueTask DisposeAsync()
+        {
+            BrowserRefreshServer?[] serversToDispose;
+
+            lock (_serversGuard)
+            {
+                serversToDispose = _servers.Values.ToArray();
+                _servers.Clear();
+            }
+
+            foreach (var server in serversToDispose)
+            {
+                if (server != null)
+                {
+                    await server.DisposeAsync();
+                }
+            }
+        }
+
+        public async ValueTask<BrowserRefreshServer?> LaunchOrRefreshBrowserAsync(
+            ProjectGraphNode projectNode,
+            ProcessSpec processSpec,
+            EnvironmentVariablesBuilder environmentBuilder,
+            ProjectOptions projectOptions,
+            CancellationToken cancellationToken)
+        {
+            // Attach trigger to the process that launches browser on URL found in the process output:
+            processSpec.OnOutput += GetBrowserLaunchTrigger(projectNode, projectOptions, cancellationToken);
```

Comment:
Is there a need to -= GetBrowserLaunchTarget somewhere?

Reply:
I don't think so - the processSpec is used to launch the process. Then it's thrown away.